File: 1358.number-of-substrings-containing-all-three-characters.py
#
# @lc app=leetcode id=1358 lang=python3
#
# [1358] Number of Substrings Containing All Three Characters
#

# @lc code=start
from collections import Counter
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        left, ans = 0, 0
        hashmap = Counter()
        for i in range(len(s)):
            hashmap[s[i]] += 1

            while len(hashmap) == 3:
                hashmap[s[left]] -= 1
                if hashmap[s[left]] == 0:
                    del hashmap[s[left]]
                left += 1

            ans += left
        return ans
# @lc code=end

File: 1358.number-of-substrings-containing-all-three-characters.java
/*
 * @lc app=leetcode id=1358 lang=java
 *
 * [1358] Number of Substrings Containing All Three Characters
 */

// @lc code=start
class Solution {
    public int numberOfSubstrings(String s) {
        int[] cnt = new int[3];
        int empty = cnt.length;
        int ans = 0;
        int left = 0;
        char[] c = s.toCharArray();
        for(int i = 0; i < c.length; i++){
            char cur = c[i];
            if(cnt[cur - 'a'] == 0){
                empty--;
            }
            cnt[cur - 'a']++;
            while(empty == 0){
                cnt[c[left] - 'a']--;
                if(cnt[c[left] - 'a'] == 0){
                    empty++;
                }
                left++;
            }
            //直接加上left就行了
            ans += left;

        }
        return ans;
    }
}
// @lc code=end

